### PR TITLE
initial testing script to evaluate page links

### PR DESCRIPTION
**Pull request by kbhawkey**
Created a script to evaluate the intra-site links on localized pages and report results.
General ideas:
- Check for localized versions of link.
- If no localized page, check and validate `en` link.
- Thinking about updating the links in the file if a localized link exists.
Need to extend check to test links containing page anchors.
Initial testing on the docs `concepts` directory.

### CODE REVIEW

**Comment by kbhawkey** [comment]
@tengqm,
Just started on this. Would it be helpful to evaluate (and possibly update) the page intra-site links?

**Comment by kbhawkey** [comment]
Thanks for the feedback @sftim.
I will make the file executable.
Definitely need to improve the reporting/diagnostics.
More things to consider.

**Comment by tengqm** [comment]
will review this in a day or two.

File: scripts/find-links-localize.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            if locale_file.find("#") >= 0:
+                locale_file = locale_file.split("#")[0]
+
+            locale_file = locale_file + ".md"
+            if Path(locale_file).is_file():
+                page_and_links.append("\nVALID en link: \n")
+                page_and_links.append(locale_file)
+            else:
+                page_and_links.append("\nNo localized file and INVALID en link:\n")
+                page_and_links.append(locale_file + "\n")
```
Comment (tengqm) [inline scripts/find-links-localize.py:113]: Some problems I pointed out when attempted to do this thoroughly in another PR (https://github.com/kubernetes/website/pull/19526#issuecomment-600938171).

- You will be treating links to redirect entries as invalid entry, e.g. `/docs/user-guide/`.
- You will have to find out whether `/docs/concepts/overview/` is a valid link, where `overview` is a directory and link target is `/docs/concepts/overview/_index.md`.
- Need to figure out that `/docs/concepts/#kubernetes-control-plane` actually points to `/docs/concepts/_index.md#kubernetes-control-plane`.
- Need to handle link target that contain hugoshort codes
- Need to handle out-of-band link targers such a `[kubenet][kubenet]`

Comment (kbhawkey) [reply scripts/find-links-localize.py:113]: @tengqm , thanks for the review. 
Yes, there will be issues with links that have redirects as well as the cases that you note.
I am thinking about checking the _redirect file for links and excluding link targets that contain shortcodes.
What are "out-of-band" link targets, [kubenet][kubenet]?
The links that include in-page anchors will be more complicated to handle.
Would like to start with one main problem (typical links that are invalid and/or have _redirects). Generally, links
to <directory>/_index.md do not change that often (are most of these "index" pages translated).
Is it worthwhile to create a tool to investigate invalid links in .md files?

Comment (tengqm) [reply scripts/find-links-localize.py:113]: @kbhawkey I believe testing tools can help diagnose quite some problems, and they are in general valuable for identifying nits that skip human eyes. I'm not against the checking or the tools. I just wanted to list some not so uncommon cases for the tools to handle. At the end of the day, the tool may be still very useful if it helps avoid some (even if not all) kinds of link errors.

Maybe I was using the incorrect term, but when I said "out-of-band" links, I was referring to cases like this:

```
The provisioner can be a [kubenet][kubenet] or a [cni plugin][cni-plugin]...

[kubenet]: https://somewhere/you/can/find/kubenet/details
[cni-plugin]: https://another/url/for/cni/specification/or/docs
```

**Comment by kbhawkey** [comment]
I plan to continue to work on this depending if this direction is correct.

**Comment by tengqm** [comment]
Ya. +2
I'd like to contribute to this as well.

**Comment by kbhawkey** [comment]
I'm placing a hold on this as it is a preliminary sketch.
@tengqm , let me know if you need help. Thanks.
/hold

File: scripts/find-links-localize.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+import sys
+import tempfile
+from pathlib import Path
+
+DIRS = ["concepts", "contribute", "home", "reference", "setup", "tasks", "tutorials"]
+TEST_DIRS = ["concepts"]
+tmp_file_output = {}
+
+def write_to_tmp():
+    """Write validation results to a temporary file."""
```
Comment (irvifa) [inline scripts/find-links-localize.py:26]: Will we report this in the CI?

File: scripts/find-links-localize.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import re
+import sys
+import tempfile
+from pathlib import Path
+
+DIRS = ["concepts", "contribute", "home", "reference", "setup", "tasks", "tutorials"]
+TEST_DIRS = ["concepts"]
+tmp_file_output = {}
+
+def write_to_tmp():
```
Comment (irvifa) [inline scripts/find-links-localize.py:25]:
```suggestion
def write_validation_result():
```

**Comment by irvifa** [comment]
For functionalities other than main if we can do some kind of unit test it will be helpful eg, expecting input and output for the test

File: scripts/find-links-localize.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    # remote_regex = re.compile(r"\]\(\/docs[\/\w\-\#\/]+\)")
+
+    # Single results: searches for pattern: []()
+    remote_regex = re.compile(r"(\[[\w\s\.-]*\])(\(\/docs[\/\w\-\#\/]+\))")
+    matches = remote_regex.findall(content)
+
+    for m in matches:
+        page_and_links.append("\nMatch link:" + m[0] + m[1])
+        # clean up path and split
+        subpaths = m[1][1:-1].split("/")
```
Comment (irvifa) [inline scripts/find-links-localize.py:76]: Probably sample of output in this part will help why we need to split it this way.

**Comment by tengqm** [comment]
@kbhawkey I was thinking if I can carry on this by pushing into your fork, :)

File: scripts/find-links-localize.py (hunk 2 above)
Comment (kbhawkey) [reply scripts/find-links-localize.py:26]: Hi @irvifa . Good questions. Thanks for your feedback.
I have not considered running this script as part of CI. This script is really an initial experimentation.

**Comment by kbhawkey** [comment]
> 
> 
> @kbhawkey I was thinking if I can carry on this by pushing into your fork, :)

Hi @tengqm . Whatever you think. I am happy to hand this over to you, or feel free to continue working with this branch.

**Comment by tengqm** [comment]
@kbhawkey I'll be working on this in the coming days and get back to team soon.

**Comment by kbhawkey** [comment]
> 
> 
> @kbhawkey I'll be working on this in the coming days and get back to team soon.

👍

**Comment by kbhawkey** [comment]
See pull request #22541 .
/close

**Pull request closed without merging**